File: mobile/src/main/java/com/siliconlabs/bledemo/adapters/ConnectionsAdapter.java
package com.siliconlabs.bledemo.adapters;

import android.bluetooth.BluetoothDevice;
import android.content.Context;

import androidx.annotation.NonNull;
import androidx.core.content.ContextCompat;
import androidx.recyclerview.widget.RecyclerView;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.LinearLayout;
import android.widget.ProgressBar;
import android.widget.TextView;

import com.siliconlabs.bledemo.R;
import com.siliconlabs.bledemo.ble.BluetoothDeviceInfo;
import com.siliconlabs.bledemo.interfaces.ServicesConnectionsCallback;
import com.siliconlabs.bledemo.utils.Constants;

import java.util.List;

public class ConnectionsAdapter extends RecyclerView.Adapter<ConnectionsAdapter.ConnectionViewHolder> {
    private List<BluetoothDevice> connectionsList;
    private Context context;
    private ServicesConnectionsCallback servicesConnectionsCallback;
    private String selectedDevice;

    public class ConnectionViewHolder extends RecyclerView.ViewHolder {
        TextView nameTV;
        Button disconnectB;
        LinearLayout layout;
        ProgressBar progressBar;

        public ConnectionViewHolder(View itemView) {
            super(itemView);
            nameTV = itemView.findViewById(R.id.textview_device_name);
            disconnectB = itemView.findViewById(R.id.button_disconnect);
            layout = itemView.findViewById(R.id.connected_device_item);
            progressBar = itemView.findViewById(R.id.progress_bar_connections);
        }

    }

    public ConnectionsAdapter(List<BluetoothDevice> bluetoothDevices, Context context) {
        this.context = context;
        this.connectionsList = bluetoothDevices;
    }

    @Override
    public void onBindViewHolder(final ConnectionsAdapter.ConnectionViewHolder holder, final int position) {
        holder.progressBar.setVisibility(View.GONE);
        holder.disconnectB.setVisibility(View.VISIBLE);
        String name = connectionsList.get(position).getName();
        String address = connectionsList.get(position).getAddress();
        if (name == null || name.equals("")) {
            name = Constants.UNKNOWN;
        }
        if (getSelectedDevice() != null) {
            if (address.toLowerCase().equals(getSelectedDevice().toLowerCase())) {
                holder.nameTV.setTextColor(ContextCompat.getColor(context,R.color.silabs_blue));
            } else {
                holder.nameTV.setTextColor(ContextCompat.getColor(context,R.color.silabs_primary_text));
            }
        } else {
            holder.nameTV.setTextColor(ContextCompat.getColor(context,R.color.silabs_primary_text));
        }
        holder.nameTV.setText(name);

        holder.disconnectB.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                BluetoothDeviceInfo bluetoothDeviceInfo = new BluetoothDeviceInfo();
                bluetoothDeviceInfo.device = connectionsList.get(position);
                servicesConnectionsCallback.onDisconnectClicked(bluetoothDeviceInfo);
            }
        });

        holder.layout.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                holder.progressBar.setVisibility(View.VISIBLE);
                holder.disconnectB.setVisibility(View.GONE);
                BluetoothDeviceInfo bluetoothDeviceInfo = new BluetoothDeviceInfo();
                bluetoothDeviceInfo.device = connectionsList.get(position);
                servicesConnectionsCallback.onDeviceClicked(bluetoothDeviceInfo);
            }
        });
    }

    @Override
    public ConnectionsAdapter.ConnectionViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.connection_item, parent, false);
        return new ConnectionViewHolder(view);
    }

    @Override
    public int getItemCount() {
        return connectionsList.size();
    }

    public void setConnectionsList(List<BluetoothDevice> connectionsList) {
        this.connectionsList = connectionsList;
    }

    public List<BluetoothDevice> getConnectionsList() {
        return connectionsList;
    }

    public void setServicesConnectionsCallback(ServicesConnectionsCallback servicesConnectionsCallback) {
        this.servicesConnectionsCallback = servicesConnectionsCallback;
    }

    public String getSelectedDevice() {
        return selectedDevice;
    }

    public void setSelectedDevice(String selectedDevice) {
        this.selectedDevice = selectedDevice;
    }
}

File: mobile/src/main/java/com/siliconlabs/bledemo/Browser/Adapters/ConnectionsAdapter.kt
package com.siliconlabs.bledemo.Browser.Adapters

import android.bluetooth.BluetoothDevice
import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.LinearLayout
import android.widget.ProgressBar
import android.widget.TextView
import androidx.core.content.ContextCompat
import androidx.recyclerview.widget.RecyclerView
import com.siliconlabs.bledemo.Bluetooth.BLE.BluetoothDeviceInfo
import com.siliconlabs.bledemo.Browser.Adapters.ConnectionsAdapter.ConnectionViewHolder
import com.siliconlabs.bledemo.Browser.ServicesConnectionsCallback
import com.siliconlabs.bledemo.R
import com.siliconlabs.bledemo.Utils.Constants
import java.util.*

class ConnectionsAdapter(var connectionsList: List<BluetoothDevice>, private val context: Context) : RecyclerView.Adapter<ConnectionViewHolder>() {
    private var servicesConnectionsCallback: ServicesConnectionsCallback? = null
    var selectedDevice: String? = null

    inner class ConnectionViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        private var tvName = itemView.findViewById(R.id.tv_device_name) as TextView
        private var btnDisconnect = itemView.findViewById(R.id.btn_disconnect) as Button
        private var progressBar = itemView.findViewById(R.id.progress_bar_connections) as ProgressBar
        var layout = itemView.findViewById(R.id.connected_device_item) as LinearLayout

        fun bind(device: BluetoothDevice) {
            progressBar.visibility = View.GONE
            btnDisconnect.visibility = View.VISIBLE

            var name = device.name
            val address = device.address
            if (name == null || name == "") {
                name = Constants.NA
            }

            if (selectedDevice != null) {
                if (address.toLowerCase(Locale.getDefault()) == selectedDevice?.toLowerCase(Locale.getDefault())) {
                    tvName.setTextColor(ContextCompat.getColor(context, R.color.silabs_blue))
                } else {
                    tvName.setTextColor(ContextCompat.getColor(context, R.color.silabs_primary_text))
                }
            } else {
                tvName.setTextColor(ContextCompat.getColor(context, R.color.silabs_primary_text))
            }

            tvName.text = name
            btnDisconnect.setOnClickListener {
                val bluetoothDeviceInfo = BluetoothDeviceInfo()
                bluetoothDeviceInfo.device = device
                servicesConnectionsCallback?.onDisconnectClicked(bluetoothDeviceInfo)
            }

            layout.setOnClickListener {
                progressBar.visibility = View.VISIBLE
                btnDisconnect.visibility = View.GONE
                val bluetoothDeviceInfo = BluetoothDeviceInfo()
                bluetoothDeviceInfo.device = device
                servicesConnectionsCallback?.onDeviceClicked(bluetoothDeviceInfo)
            }

        }
    }

    override fun onBindViewHolder(holder: ConnectionViewHolder, position: Int) {
        val device = connectionsList[position]
        holder.bind(device)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ConnectionViewHolder {
        val view = LayoutInflater.from(parent.context).inflate(R.layout.adapter_connection, parent, false)
        return ConnectionViewHolder(view)
    }

    override fun getItemCount(): Int {
        return connectionsList.size
    }

    fun setServicesConnectionsCallback(servicesConnectionsCallback: ServicesConnectionsCallback?) {
        this.servicesConnectionsCallback = servicesConnectionsCallback
    }

}